libraries/Adapters/Microsoft.Bot.Builder.Adapters.Twilio/TwilioAdapter.cs

```diff
@@ -0,0 +1,243 @@
+        public override async Task<ResourceResponse> UpdateActivityAsync(ITurnContext turnContext, Activity activity, CancellationToken cancellationToken)
+        {
+            // Twilio adapter does not support updateActivity.
+            return await Task.FromException<ResourceResponse>(new NotSupportedException("Twilio SMS does not support updating activities."));
+        }
+
+        /// <summary>
+        /// Standard BotBuilder adapter method to delete a previous message.
+        /// </summary>
+        /// <param name="turnContext">A TurnContext representing the current incoming message and environment.</param>
+        /// <param name="reference">An object in the form "{activityId: `id of message to delete`, conversation: { id: `id of channel`}}".</param>
+        /// <param name="cancellationToken">A cancellation token for the task.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public override async Task DeleteActivityAsync(ITurnContext turnContext, ConversationReference reference, CancellationToken cancellationToken)
+        {
+            // Twilio adapter does not support deleteActivity.
+            await Task.FromException<ResourceResponse>(new NotSupportedException("Twilio SMS does not support deleting activities."));
+        }
+
+        /// <summary>
+        /// Formats a BotBuilder activity into an outgoing Twilio SMS message.
+        /// </summary>
+        /// <param name="activity">A BotBuilder Activity object.</param>
+        /// <returns>A Message's options object with {body, from, to, mediaUrl}.</returns>
+        private CreateMessageOptions ActivityToTwilio(Activity activity)
+        {
+            var mediaUrls = new List<Uri>();
+
+            if ((activity.ChannelData as dynamic)?.mediaURL != null)
+            {
+                mediaUrls.Add(new Uri((activity.ChannelData as dynamic).mediaURL));
+            }
+
+            var messageOptions = new CreateMessageOptions(activity.Conversation.Id)
+            {
+                ApplicationSid = activity.Conversation.Id,
+                From = _options.TwilioNumber,
+                Body = activity.Text,
+                MediaUrl = mediaUrls,
+            };
+
+            return messageOptions;
+        }
+
+        /// <summary>
+        /// Standard BotBuilder adapter method for continuing an existing conversation based on a conversation reference.
+        /// </summary>
+        /// <param name="reference">A conversation reference to be applied to future messages.</param>
+        /// <param name="logic">A bot logic function that will perform continuing action in the form 'async(context) => { ... }'.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        private async Task ContinueConversationAsync(ConversationReference reference, BotCallbackHandler logic)
```

Comment:
This method doesn't seem to be referenced anywhere in the code, do we need it?

Reply:
We turned this method public. It will be used by the user.